opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py

```diff
@@ -145,6 +145,70 @@ def from_map(cls, maxlen, mapping):
 Link = namedtuple("Link", ("context", "attributes"))
 
 
+class SpanProcessor:
+    """Interface which allows hooks for SDK's `Span`s start and end method
+    invocations.
+
+    Span processors can be registered directly using
+    :func:`~Tracer:add_span_processor` and they are invoked in the same order
+    as they were registered.
+    """
+
+    def on_start(self, span: "Span") -> None:
+        """Called when a :class:`Span` is started.
+
+        This method is called synchronously on the thread that starts the
+        span, therefore it should not block or throw an exception.
+
+        Args:
+            span: The :class:`Span` that just started.
+        """
+
+    def on_end(self, span: "Span") -> None:
+        """Called when a :class:`Span` is ended.
+
+        This method is called synchronously on the thread that ends the
+        span, therefore it should not block or throw an exception.
+
+        Args:
+            span: The :class:`Span` that just ended.
+        """
+
+    def shutdown(self) -> None:
+        """Called when a :class:`Tracer` is shutdown."""
+
+
+# Used when no SpanProcessor has been added to the Tracer
+_NO_OP_SPAN_PROCESSOR = SpanProcessor()
+
+
+class MultiSpanProcessor(SpanProcessor):
```

Comment:
Is there a motivation to have MultiSpanProcessor be a specialized version of SpanProcessor? In general I've thought if an interface probably needs to support adding multiple processors, it would just be better to have the system consuming the SpanProcessor to take multiple?

e.g.
```
tracer.set_span_processor(Processor1(), Processor2())
```

rather than

```
tracer.set_span_processor(MultiSpanProcessor(Processor1(), Processor2())
```
```

Reply:
I think this use of the [Composite Pattern](https://en.wikipedia.org/wiki/Composite_pattern) is OK, as it is IMHO a well-understood coding-pattern. Putting that code directly in the span would not be much simpler. Though I agree that it currently *is* simpler and I don't know if we will ever have more consumers of span processors (though it is imaginable to add processors for before and after sampling stages, for example)